SeriesCruncher/app/src/main/java/com/balazs_csernai/seriescruncher/utils/ui/color/ColorProviderImpl.java

```diff
@@ -72,34 +76,46 @@ private Swatch findSecondarySwatch(@NonNull List<Swatch> swatches) {
         return swatch;
     }
 
-    @ColorInt
-    @Override
-    public int getPrimaryTextColor() {
-        if (calculateColorBrightness(getPrimaryBackgroundColor()) >= BRIGHTNESS_THRESHOLD) {
-            return getColor(R.color.black);
+    private boolean areColorsDifferent(Swatch swatchOne, Swatch swatchTwo) {
+        return COLOR_THRESHOLD < calculateColorDifference(swatchOne.getRgb(), swatchTwo.getRgb());
+    }
+
+    private int calculateColorDifference(@ColorInt int colorOne, @ColorInt int colorTwo) {
+        return Math.abs(Color.red(colorOne) - Color.red(colorTwo))
+                + Math.abs(Color.green(colorOne) - Color.green(colorTwo))
+                + Math.abs(Color.blue(colorOne) - Color.blue(colorTwo));
+    }
+
+    private Swatch findAccentSwatch(@NonNull List<Swatch> swatches) {
+        swatches.remove(primarySwatch);
+        swatches.remove(secondarySwatch);
+        if (swatches.isEmpty()) {
+            return null;
         } else {
-            return getColor(R.color.white);
+            return swatches.get(0);
         }
     }
 
-    @ColorInt
     @Override
-    public int getPrimaryBackgroundColor() {
-        return primarySwatch == null ? getColor(R.color.colorPrimary) : primarySwatch.getRgb();
+    public ColorModel getPrimaryColor() {
+        return createColorModel(primarySwatch, R.color.colorPrimary);
     }
 
     @Override
-    public int getSecondaryTextColor() {
-        if (calculateColorBrightness(getSecondaryBackgroundColor()) >= BRIGHTNESS_THRESHOLD) {
-            return getColor(R.color.black);
-        } else {
-            return getColor(R.color.white);
-        }
+    public ColorModel getSecondaryColor() {
+        return createColorModel(secondarySwatch, R.color.colorSecondary);
     }
 
     @Override
-    public int getSecondaryBackgroundColor() {
-        return secondarySwatch == null ? getColor(R.color.colorAccent) : secondarySwatch.getRgb();
+    public ColorModel getAccentColor() {
+        @ColorInt int bg = accentSwatch == null ? getColor(R.color.colorAccent) : accentSwatch.getRgb();
+        @ColorInt int fg = accentSwatch == null ? calculateForegroundColor(bg) : accentSwatch.getBodyTextColor();
+        return new ColorEntity(bg, fg);
+    }
+
+    private ColorModel createColorModel(Swatch swatch, @ColorRes int defaultColorRes) {
```

Comment:
We should unify color generation and use bodyTextColor for primary and secondary swatches also.